src/server/auth/server/api_server.go

```diff
@@ -1416,22 +1368,147 @@ func (a *apiServer) RestoreAuthToken(ctx context.Context, req *auth.RestoreAuthT
+func (a *apiServer) lookupAuthTokenTTLSeconds(ctx context.Context, tokenHash string) (int64, error) {
+	var ttl float64
+	if err := a.env.GetDBClient().GetContext(ctx, &ttl,
+		`SELECT ROUND( EXTRACT( EPOCH FROM (expiration - NOW()))) as ttl 
+		FROM auth.auth_tokens WHERE token_hash = $1 AND expiration IS NOT NULL`, tokenHash); err != nil {
+		if err == sql.ErrNoRows {
+			return -1, nil
+		}
+		return 0, err
+	}
+	return int64(ttl), nil
+}
+
+func (a *apiServer) lookupAuthTokenInfo(ctx context.Context, tokenHash string) (*auth.TokenInfo, error) {
+	var tokenInfo auth.TokenInfo
+	if err := a.env.GetDBClient().GetContext(ctx, &tokenInfo, `SELECT subject FROM auth.auth_tokens WHERE token_hash = $1`, tokenHash); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, col.ErrNotFound{Type: "auth_tokens", Key: tokenHash}
+		}
+		return nil, err
+	}
+	return &tokenInfo, nil
+}
+
+func (a *apiServer) listRobotTokens(ctx context.Context) ([]*auth.HashedAuthToken, error) {
+	rows, err := a.env.GetDBClient().QueryxContext(ctx,
+		`SELECT token_hash as tokenHash, subject, ROUND( EXTRACT( EPOCH FROM (expiration - NOW()))) as ttl  
+		FROM auth.auth_tokens 
+		WHERE subject LIKE $1 || '%'`, auth.RobotPrefix)
+	if err != nil {
+		return nil, err
+	}
+	robotTokens := make([]*auth.HashedAuthToken, 0)
+	for rows.Next() {
+		var tokenHash string
+		var subject string
+		var ttl *float64
+		if err = rows.Scan(&tokenHash, &subject, &ttl); err != nil {
+			return nil, errors.Wrapf(err, "error querying token")
+		}
+
+		token := &auth.HashedAuthToken{
+			HashedToken: tokenHash,
+			TokenInfo: &auth.TokenInfo{
+				Subject: subject,
+			},
+		}
+		if ttl != nil {
+			expiration, err := types.TimestampProto(time.Now().Add(time.Second * time.Duration(int64(*ttl))))
+			if err != nil {
+				return nil, err
+			}
+			token.Expiration = expiration
+		}
+		robotTokens = append(robotTokens, token)
+	}
+	return robotTokens, nil
+}
+
+func (a *apiServer) generateAndInsertAuthToken(ctx context.Context, subject string, ttlSeconds int64) (string, error) {
+	token := uuid.NewWithoutDashes()
+	if err := a.upsertAuthToken(ctx, auth.HashToken(token), subject, ttlSeconds); err != nil {
+		return "", err
+	}
+	return token, nil
+}
+
+func (a *apiServer) generateAndInsertAuthTokenNoTTL(ctx context.Context, subject string) (string, error) {
+	token := uuid.NewWithoutDashes()
+	if err := a.upsertAuthTokenNoTTL(ctx, auth.HashToken(token), subject); err != nil {
+		return "", err
+	}
+
+	return token, nil
+}
+
+// generates a token, and stores it's hash and supporting data in postgres
+func (a *apiServer) upsertAuthToken(ctx context.Context, tokenHash string, subject string, ttlSeconds int64) error {
+	// Register the pachd in the database
+	if _, err := a.env.GetDBClient().ExecContext(ctx,
+		`INSERT INTO auth.auth_tokens (token_hash, subject, expiration) 
+		VALUES ($1, $2, NOW() + $3 * interval '1 sec') 
+		ON CONFLICT (token_hash) 
```

Comment:
What's the use case for updating the subject and expiration if the hash collides? I do think we used to do this, but if a user is going to get a token that exactly matches another user's token we should probably throw an error?

Reply:
I think it was needed in the GetAuthToken RPC that's now removed. I will simplify it here to a plain insert